src/Common/src/Internal/Runtime/CompilerHelpers/StartupCodeHelpers.cs

```diff
@@ -144,16 +146,41 @@ private static unsafe void RunEagerClassConstructors(IntPtr cctorTableStart, int
         private static unsafe void InitializeStatics(IntPtr gcStaticRegionStart, int length)
         {
             IntPtr gcStaticRegionEnd = (IntPtr)((byte*)gcStaticRegionStart + length);
+
             for (IntPtr* block = (IntPtr*)gcStaticRegionStart; block < (IntPtr*)gcStaticRegionEnd; block++)
             {
                 // Gc Static regions can be shared by modules linked together during compilation. To ensure each
                 // is initialized once, the static region pointer is stored with lowest bit set in the image.
                 // The first time we initialize the static region its pointer is replaced with an object reference
                 // whose lowest bit is no longer set.
+                // Note that this also properly skips the optional preinit data pointer since it is pointer aligned, 
+                // when this entry is already initialized.
                 IntPtr* pBlock = (IntPtr*)*block;
-                if (((*pBlock).ToInt64() & 0x1L) == 1)
+                long blockAddr = (*pBlock).ToInt64();
+                if ((blockAddr & 0x1L) == 1)
                 {
-                    object obj = RuntimeImports.RhNewObject(new EETypePtr(new IntPtr((*pBlock).ToInt64() & ~0x1L)));
+                    object obj = RuntimeImports.RhNewObject(new EETypePtr(new IntPtr(blockAddr & ~0x3L)));
+
+                    if ((blockAddr & 0x2L) == 2)
+                    {
+                        // The next pointer is preinitialized data blob that contains preinitialized static GC fields,
+                        // which are pointer relocs to GC objects in frozen segment. 
+                        // It actually has all GC fields including non-preinitialized fields and we simply copy over the
+                        // entire blob to this object, overwriting everything. 
+                        fixed (IntPtr* pEEType = &obj.m_pEEType)
+                        {
+                            // Skip SyncBlock + EEType when calculating field size
+                            int fieldSize = (int)obj.EETypePtr.BaseSize - IntPtr.Size * 2;
```

Comment:
> I'm wondering if this should be an API on EETypePtr.

On a second thought, we might want to make this an internal method on `System.Object` itself. This complements the existing `GetRawData` method on `System.Object`: it gives you the *size of the raw data*. I would call this `GetRawDataSize()`. That way we avoid hardcoding assumptions about System.Object's layout in here. which just feels dirty.